### ISSUE TITLE
[Java][Feature] Generating operation methods used to build a graph

### ISSUE THREAD

**Issue by karllessard** [tensorflow/tensorflow#7149]
As suggested in https://www.tensorflow.org/versions/r0.11/how_tos/language_bindings, the list of operation methods for building a graph should be generated dynamically from the list exposed by the core, using preferably protobuf. 

Is anyone already working on that feature? If not, I'm tempted to try it. I was thinking of generating a builder hierarchy that allows to optionally set an argument after adding an operation. For example
```java
GraphBuider
    .matMul(a, b)
        .withTransposeB(true)
    .softmax(logits)
    ...
```
and the GraphBuilder classes would be generated at build time by Bazel and would make use of the already existing OperationBuilder. Since I'm new to Tensorflow, please tell me if that doesn't make any sense to you

Thanks

(p.s. as suggested by @drpngx, I've started this discussion as a seperate issue to avoid continuing the #5 saga)

**Comment by drpngx** [comment]
Thanks @karllessard! @asimshankar should know if that's in the plans or if it would be a welcome contribution.

**Comment by asimshankar** [comment]
Having generated code for ops in Java is indeed a next step for the Java API, but we haven't gotten around to it, or figuring out what an appropriate API would be.

We'd be happy to hear any suggestions/proposals you may have. As for the suggestion above, I don't think we want to put all operations as methods in a single `GraphBuilder` class for a few reasons such as (a) Per-op attribute handling seems a bit tricky, (b) What about custom operations defined by other external libraries (e.g. as per the [Adding a new op HOWTO](https://www.tensorflow.org/how_tos/adding_an_op/)).

We can take inspiration from the generated C++ API (see the [label_image example](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/label_image/main.cc#L102) and the Go API, but it does require more thought.

**Comment by karllessard** [comment]
I agree that exposing operations in their own particular object would create a richer API than declaring everything into a single builder class, I'll also get my mind into it.

Now, without insisting, I'll clarify how my suggestion would have handled point a), maybe some of those ideas can morph into something else:

1. The generated builder class implements each operations by returning a specialized builder that inherits this same class (e.g. GraphBuilder.matMul() returns a MatMulGraphBuilder that inherits GraphBuilder).

2. The specialized builder exposes all other operations directly from its superclass (for going on with graph building) + methods for setting optional attributes of the current operation, prefixed by "with" (e.g. withTransposeA() and withTransposeB()).

3. The specialized builder is responsible of building the current operation, using internally an OperationBuilder.

4. Invoking another operation on a specialized builder will therefore end the construction of the current operation and return another specialized builder for building the new operation.

**Comment by karllessard** [comment]
Anyone curious to see how a static API for building a graph may look like, I've pushed a little POC to my [repo](https://github.com/karllessard/tensorflow/tree/java-static-api/tensorflow/java) (this [commit](https://github.com/karllessard/tensorflow/commit/d2a5518a5830ec9d0b81912b95e3e33b64bd5ec5))  that minimally fulfill the requirements of the [LabelImage](https://github.com/karllessard/tensorflow/blob/java-static-api/tensorflow/java/src/main/java/org/tensorflow/examples/LabelImage2.java) example.

This POC focuses on the following point:

1. Group operations into related topics, like the C++ client does, by generating a static API per /tensorflow/core/ops/* library (e.g. [ImageOps](https://github.com/karllessard/tensorflow/blob/java-static-api/tensorflow/java/src/main/java/org/tensorflow/ops/image/ImageOps.java), [ArrayOps](https://github.com/karllessard/tensorflow/blob/java-static-api/tensorflow/java/src/main/java/org/tensorflow/ops/array/ArrayOps.java), ...)

2. Allow optional parametrization using operation-specific methods encapsulated in their own classes (e.g. the [DecodeJpeg](https://github.com/karllessard/tensorflow/blob/java-static-api/tensorflow/java/src/main/java/org/tensorflow/ops/image/DecodeJpeg.java) interface)

3. Keep the simplicity and elegance of the solution from the previous version of LabelImage while allowing more flexibility to support all cases (e.g. multi-output operations, etc). The following lists supported technics to connect nodes together.
  
a) Functionally
```java
op1(s, op2(s, const(s, x)), const(s, y));
``` 
b) By node references
```java
Const c1 = const(s, x);
Const c2 = const(s, y);
op1(s, op2(s, c1), c2);
```
c) By node lookups
```java
op2(s, "op2", const(s, x));
op1(s, s.node("op2"), const(s, y));
```

Technics b) and c) allows us to access output value other than the default (0) one for multi-output operations. For example:
```java
op1(s, s.node("op2").output(1), const(s, y));
```

Any comments are more than welcomed

**Comment by kbsriram** [comment]
@karllessard and I (with guidance from @asimshankar) have consolidated ideas on this thread into [a proposal for the API](https://docs.google.com/document/d/1kJ1lak4OIz0b7N6am3PSRe6i1enZgRHjbj8U-8PJ68w/edit?usp=sharing). We'd love to hear your feedback, so please continue to add comments on this issue.

We'd also like to help move things forward via code contributions, and Karl and I plan to start some related PRs. Please let us know if you're interested/working on something related, so we can jointly work to bring Java bindings for Tensorflow operators.

**Comment by michaelisard** [comment]
Marking 'contributions welcome' since this is an issue actively being worked on.

**Comment by Moriadry** [comment]
I saw so many ops with plenty of many properties and methods, such as array_ops, control_flow_ops, nn_ops and so on,  in c++ api. Does that mean we will implement them all in Java api?

**Comment by karllessard** [comment]
Yes, they will all be automatically implemented in Java. All operations found in the C++ core are to be wrapped by classes generated at compile-time (TensorFlow-speaking) that will expose their inputs, attributes and outputs with named variables and methods.

**Comment by HedgehogCode** [comment]
What is the current status of this issue? Is someone working on it and is it supposed to be resolved soon or is it a goal which lies far in the future?

**Comment by karllessard** [comment]
It is still ongoing and actively being committed but it takes some time, there is a lot of code and iterations involved. Pull request #14094 is related to this story and is currently being reviewed by @asimshankar .
